homeassistant/components/environment_canada/sensor.py

```diff
@@ -49,107 +51,98 @@ def validate_station(station):
+        """Fetch data from Environment Canada."""
+        async with async_timeout.timeout(10):
+            await ec_data.update()
+        ec_data.conditions.update(ec_data.alerts)
+        return ec_data.conditions
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="environment_canada_sensor",
+        update_method=async_update_data,
+        update_interval=timedelta(minutes=5),
+    )
 
-    sensor_list = list(ec_data.conditions) + list(ec_data.alerts)
-    add_entities([ECSensor(sensor_type, ec_data) for sensor_type in sensor_list], True)
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
 
+    async_add_devices(
+        ECSensor(coordinator, sensor_type, ec_data.metadata)
+        for sensor_type in coordinator.data.keys()
+    )
 
-class ECSensor(Entity):
+
+class ECSensor(CoordinatorEntity, Entity):
     """Implementation of an Environment Canada sensor."""
 
-    def __init__(self, sensor_type, ec_data):
+    def __init__(self, coordinator, sensor_type, metadata):
         """Initialize the sensor."""
+        super().__init__(coordinator)
         self.sensor_type = sensor_type
-        self.ec_data = ec_data
-
-        self._unique_id = None
-        self._name = None
-        self._state = None
-        self._attr = None
-        self._unit = None
+        self.metadata = metadata
 
     @property
     def unique_id(self) -> str:
         """Return the unique ID of the sensor."""
-        return self._unique_id
+        return f"{self.metadata['location']}-{self.sensor_type}"
 
     @property
     def name(self):
         """Return the name of the sensor."""
-        return self._name
+        return self.coordinator.data[self.sensor_type].get("label")
 
     @property
     def state(self):
         """Return the state of the sensor."""
-        return self._state
-
-    @property
-    def device_state_attributes(self):
-        """Return the state attributes of the device."""
-        return self._attr
-
-    @property
-    def unit_of_measurement(self):
-        """Return the units of measurement."""
-        return self._unit
-
-    def update(self):
-        """Update current conditions."""
-        self.ec_data.update()
-        self.ec_data.conditions.update(self.ec_data.alerts)
-
-        conditions = self.ec_data.conditions
-        metadata = self.ec_data.metadata
-        sensor_data = conditions.get(self.sensor_type)
-
-        self._unique_id = f"{metadata['location']}-{self.sensor_type}"
-        self._attr = {}
-        self._name = sensor_data.get("label")
-        value = sensor_data.get("value")
+        value = self.coordinator.data[self.sensor_type].get("value")
 
         if isinstance(value, list):
-            self._state = " | ".join([str(s.get("title")) for s in value])[:255]
-            self._attr.update(
-                {ATTR_TIME: " | ".join([str(s.get("date")) for s in value])}
```

Comment:
Looks like we're removing this state attribute. Is that intentional? That would be a breaking change.

Reply:
Good point, I was able to work this back in.